src/smart-escrow/SmartEscrow.sol

```diff
@@ -0,0 +1,239 @@
+     * @param startTimestamp Timestamp of the start of vesting period (or the cliff, if there is one).
+     * @param endTimestamp Timestamp of the end of the vesting period.
+     * @param vestingPeriodSeconds Period of time between each vesting event in seconds.
+     * @param numInitialTokens Number of OP tokens which vest at start time.
+     * @param numVestingEventTokens Number of OP tokens which vest upon each vesting event.
+     */
+    constructor(
+        address beneficiaryOwner,
+        address benefactorOwner,
+        address escrowOwner,
+        address beneficiaryAddress,
+        address benefactorAddress,
+        uint64 startTimestamp,
+        uint64 endTimestamp,
+        uint64 vestingPeriodSeconds,
+        uint64 numInitialTokens,
+        uint64 numVestingEventTokens
+    ) {
+        require(beneficiaryOwner != address(0), "SmartEscrow: beneficiaryOwner is zero address");
+        require(benefactorOwner != address(0), "SmartEscrow: benefactorOwner is zero address");
+        require(escrowOwner != address(0), "SmartEscrow: escrowOwner is zero address");
+        require(beneficiaryAddress != address(0), "SmartEscrow: beneficiaryAddress is zero address");
+        require(benefactorAddress != address(0), "SmartEscrow: benefactorAddress is zero address");
+
+        require(startTimestamp > 0, "SmartEscrow: start time is zero");
+        require(endTimestamp > startTimestamp, "SmartEscrow: end time is before start time");
+        require(vestingPeriodSeconds > 0, "SmartEscrow: vestingPeriodSeconds is zero");
+
+        // TODO: could add a check that checks the owners of the escrowOwner are beneficiaryOwner and benefactorOwner
+        _grantRole(DEFAULT_ADMIN_ROLE, escrowOwner);
+        _grantRole(ESCROW_OWNER, escrowOwner);
+        _grantRole(BENEFICIARY_OWNER, beneficiaryOwner);
+        _grantRole(BENEFACTOR_OWNER, benefactorOwner);
+
+        _beneficiary = beneficiaryAddress;
+        _benefactor = benefactorAddress;
+        _start = startTimestamp;
+        _end = endTimestamp;
+        _vestingPeriod = vestingPeriodSeconds;
+        _initialTokens = numInitialTokens;
+        _vestingEventTokens = numVestingEventTokens;
+    }
+
+    /**
+     * @dev Allow escrow owner (likely a 2-of-2 multisig with beneficiary and benefactor
+     * owners as signers) to terminate the contract.
+     *
+     * Emits a {ContractTerminated} event.
+     */
+    function terminate() external virtual onlyRole(ESCROW_OWNER) {
+        _contractTerminated = true;
+        emit ContractTerminated();
+        withdrawUnvestedTokens();
+    }
+
+    /**
+     * @dev Allow beneficiary owner to update beneficiary address
+     *
+     * Emits a {BeneficiaryUpdated} event.
+     */
+    function updateBeneficiary(address newBeneficiary) external virtual onlyRole(BENEFICIARY_OWNER) {
+        require(newBeneficiary != address(0), "SmartEscrow: newBeneficiary is zero address");
```

Comment:
nit on these errors, we had talked about migrating to custom errors as that's sort of latest convention.